### DISCUSSION TITLE
Plot a {sf} `MULTILINE` in 3D

### DISCUSSION THREAD

**Discussion by pierreroudier** [tylermorganwall/rayshader#167] [Q&A] [answered]
I am trying to replicate the (very cool) 3D plot of a line across the landscape, as pictured [in this tweet](https://twitter.com/CraigTaylorViz/status/1404855283445272578):

![](https://pbs.twimg.com/media/E38LrLoXMAktePd?format=jpg&name=4096x4096)

I don't seem to be able to properly render the `MULTILINE` geometry, though.

```
library(rayshader)
library(rayrender)

# Read example coordinates extracted from the `sf` object
source("https://gist.githubusercontent.com/pierreroudier/96d07509f2f45dbafed393fb1da08d29/raw/f63faa9aa7ba8cb9a4b0e70881653288c73d31b6/gistfile1.txt")

scenelist = list() 
counter = 1
  
for(j in 1:(nrow(coords) - 1)) {
  
  scenelist[[counter]] = segment(
    start = c(
      coords$X[j] / 100000, 
      coords$height[j] / 360,
      coords$Y[j] / 100000
      ),
    end = c(
      coords$X[j + 1] / 100000,
      coords$height[j + 1] / 360,
      coords$Y[j + 1] / 100000
      ),
    radius = 0.3,
    material = metal() 
  )
  counter = counter + 1
}

fullscene = do.call(rbind, scenelist)

generate_ground(
  material = diffuse(color="grey20")
  ) %>%
  add_object(
    fullscene
  ) %>%
  render_scene(
    lookfrom = c(0, 50, -40), # c(1821639, 5515276, 450), 
    lookat = c(0, -5, 0), # c(1821748, 5515390, -375), 
    samples = 200, 
    # fov = 0, 
    width = 1000, 
    height = 1000
  )
```

There are two steps that get me confused:

- I tried to adjust the coordinates manually, to take into account their scaling
- I am pretty sure my issue is with finding good parameters for `render_scene`

Thanks!

**Comment by tylermorganwall** [comment, marked as answer]
You're running into floating point error issues, as those numbers are extremely large—when you scale them down, your line segments are extremely small. You just need to remove the offset that's reducing your dynamic range:

``` r
source("https://gist.githubusercontent.com/pierreroudier/96d07509f2f45dbafed393fb1da08d29/raw/f63faa9aa7ba8cb9a4b0e70881653288c73d31b6/gistfile1.txt")
apply(coords,2,range)
#>            X       Y  height
#> [1,] 1821639 5515276 363.439
#> [2,] 1821868 5515481 384.405

coords$X = coords$X - min(coords$X)
coords$Y = coords$Y - min(coords$Y)


apply(coords,2,range)
#>            X        Y  height
#> [1,]   0.000   0.0000 363.439
#> [2,] 228.889 204.9561 384.405

library(rayshader)
library(rayrender)

scenelist = list() 
counter = 1

for(j in 1:(nrow(coords)-1)) {
  
  scenelist[[counter]] = segment(
    start = c(
      coords$X[j], 
      coords$height[j]/36,
      coords$Y[j]
    ),
    end = c(
      coords$X[j + 1],
      coords$height[j + 1]/36,
      coords$Y[j + 1]
    ),
    radius = 1,
    material = light(intensity=5,color="purple") 
  )
  counter = counter + 1
}

fullscene = do.call(rbind, scenelist)

generate_ground(
  material = diffuse(color="grey20")
) %>%
  add_object(
    fullscene
  ) %>%
  render_scene(
    lookfrom = c(-40, 150, -40), # c(1821639, 5515276, 450), 
    lookat = c(100, 37, 100), # c(1821748, 5515390, -375), 
    samples = 10, 
    fov = 80,
    width = 1000, 
    height = 1000,
    ambient_light = T,
    backgroundlow = "black",
    backgroundhigh = "#220000",
  )
```

![](https://i.imgur.com/mJjxqMo.png)

If you plan on adding terrain under the path, just offset the terrain by the same values and they should match.

**Reply by pierreroudier** [reply]
Thanks, it works great!

Another question that puzzles me: why does `segment` takes `start` and `end` coordinates in the `x, y, z` order, rather than `x, y, z`?

**Reply by tylermorganwall** [reply]
I assume you mean `x, z, y`: The choice of the y-axis being "up" is a computer graphics convention, and it's the standard used in most OpenGL-based software, which in R means `rgl`. Because `rayshader` uses `rgl` to drive its interactive 3D graphics, it also inherited this choice, and then `rayrender` stuck with it for consistency.

No particular reason why the up axis couldn't be `z`, but it's just a choice that one needs to make and stick with!

**Reply by pierreroudier** [reply]
Thanks -- again it's a matter of re-formatting my GIS mind, where x, y, and z are pretty much set in stone

**Reply by wiesehahn** [reply]
How would I add a terrain created with rayshader to such lines or vice versa?
